### PR TITLE
Create Cloud Run deployment branch

### PR DESCRIPTION
**Pull request by petechd**
### What is the context of this PR?
This branch is used by new Cloud Run test pipeline in [this PR](https://github.com/ONSdigital/eq-pipelines/pull/241).

### How to review 
Check if script and yaml file changes are all correct and variables are refactored so there are no hardcoded left.

### CODE REVIEW

**Comment by petechd** [comment]
[Pipeline a Cloud Run environment (3 SP)](https://trello.com/c/CjGGHDpw/4764-pipeline-a-cloud-run-environment-3-sp)

File: ci/deploy_app.sh [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+IMAGE_TAG="${IMAGE_TAG:=latest}"
+REGION="${REGION:=europe-west2}"
+MIN_INSTANCES="${MIN_INSTANCES:=1}"
+MAX_INSTANCES="${MAX_INSTANCES:=1}"
+
+gcloud beta run deploy eq-questionnaire-launcher \
+    --project="${PROJECT_ID}" --region="${REGION}" --concurrency=250 --min-instances="${MIN_INSTANCES}" --max-instances="${MAX_INSTANCES}" \
+    --port=8000 --cpu=1 --memory=256M \
+    --image="${DOCKER_REGISTRY}/eq-questionnaire-launcher:${IMAGE_TAG}" --platform=managed --allow-unauthenticated \
+    --set-env-vars SURVEY_RUNNER_URL="https://${RUNNER_URL}"
```
Comment (ajmaddaford) [inline ci/deploy_app.sh:13]: The full runner URL should be passed, rather than get concatenated with `https://` in this script.

Comment (petechd) [reply ci/deploy_app.sh]: Ok, made relevant changes in pipeline so no concatenation here now.

**Comment by MebinAbraham** [comment]
The README needs updating with these changes. They still references k8s/helm etc.

File: ci/deploy_app.sh [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+set -euxo pipefail
+
+IMAGE_TAG="${IMAGE_TAG:=latest}"
+REGION="${REGION:=europe-west2}"
+MIN_INSTANCES="${MIN_INSTANCES:=1}"
+MAX_INSTANCES="${MAX_INSTANCES:=1}"
+
+gcloud beta run deploy eq-questionnaire-launcher \
+    --project="${PROJECT_ID}" --region="${REGION}" --concurrency=250 --min-instances="${MIN_INSTANCES}" --max-instances="${MAX_INSTANCES}" \
+    --port=8000 --cpu=1 --memory=256M \
```
Comment (MebinAbraham) [inline ci/deploy_app.sh:11]: Can we make concurrency, CPU and memory also configurable but set them with defaults?

Comment (petechd) [reply ci/deploy_app.sh]: Made them configurable 👍.

**Comment by petechd** [comment]
> The README needs updating with these changes. They still reference k8s/helm etc.

Updated both README files so they are relevant now.

File: ci/README.md [hunk 3]
```diff
@@ -13,15 +13,18 @@ The following are optional variables that can also be set if needed:
...
 - REGION
 - IMAGE_TAG
+- CONCURRENCY  
+- MIN_INSTANCES
+- MAX_INSTANCES
+- CPU
+- MEMORY
 
-	To deploy the app to the cluster via Concourse, use the following task command, specifying the `image_registry` and the `deploy_image_version` variables:
+	To deploy the app to the cluster via Concourse, use the following task command:
```
Comment (ajmaddaford) [inline ci/README.md:22]:
```suggestion
To deploy the app to the cluster via Concourse, use the following task command:
```

**Review by ajmaddaford — APPROVED** [review]

File: README.md [hunk 4]
```diff
@@ -54,33 +54,9 @@ Now run Go launcher and navigate to "http://localhost:8000/quick-launch?url=" pa
...
-- RUNNER_URL
-- DOCKER_REGISTRY
-- IMAGE_TAG
-
-To deploy to a cluster you can run the following command
-
-```
-./k8s/deploy_app.sh
-```
+For deploying with Concourse see the [CI README](./ci/README.md).
```
Comment (MebinAbraham) [inline README.md:59]: As discussed, we should also document how to deploy directly via `gcloud`.

File: ci/deploy_app.sh [hunk 5]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
+set -euxo pipefail
+
```
Comment (MebinAbraham) [inline ci/deploy_app.sh:3]: The project ID and runner URL needs to be validated to ensure they exist.

File: README.md (hunk 4 above)
Comment (petechd) [reply README.md]: Documented direct deployment using `gcloud` now.

File: ci/deploy_app.sh (hunk 5 above)
Comment (petechd) [reply ci/deploy_app.sh:3]: Added variables validation now, thanks!

File: README.md [hunk 6]
```diff
@@ -54,33 +54,36 @@ Now run Go launcher and navigate to "http://localhost:8000/quick-launch?url=" pa
...
-```
-
-You need to have Helm installed locally
+### Deployment with [gcloud](https://cloud.google.com/sdk/gcloud)
 
-1. Install Helm with `brew install kubernetes-helm` and then run `helm init --client-only`
+To deploy this application with gcloud, you must be logged in using `gcloud auth login` and `gcloud auth application-default login`.
 
-1. Install Helm Tiller plugin for tillerless deploys `helm plugin install https://github.com/rimusz/helm-tiller`
+Make sure required [environmental variables](./README.md#environment-variables) are set.
```
Comment (ajmaddaford) [inline README.md:65]:
```suggestion
Make sure the required [environment variables](./README.md#environment-variables) are set.
```

File: ci/README.md [hunk 7]
```diff
@@ -1,27 +1,17 @@
...
-To deploy this application with Concourse, you must have a Kubernetes cluster and be logged in to a Concourse instance.
+To deploy this application with Concourse, you must be logged in to a Concourse instance.
 
 ---
 
-The following environment variables must be set when deploying the app:
-- PROJECT_ID
-- DOCKER_REGISTRY
-- RUNNER_URL
+Make sure required [environmental variables](../README.md#environment-variables) are set.
```
Comment (ajmaddaford) [inline ci/README.md:7]:
```suggestion
Make sure the required [environment variables](../README.md#environment-variables) are set.
```

File: ci/deploy_app.sh [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+CONCURRENCY="${CONCURRENCY=250}"
+MIN_INSTANCES="${MIN_INSTANCES:=1}"
+MAX_INSTANCES="${MAX_INSTANCES:=1}"
+CPU="${CPU=1}"
+MEMORY="${MEMORY=256M}"
+
+gcloud beta run deploy eq-questionnaire-launcher \
+    --project="${PROJECT_ID}" --region="${REGION}" --concurrency="${CONCURRENCY}" --min-instances="${MIN_INSTANCES}" --max-instances="${MAX_INSTANCES}" \
+    --port=8000 --cpu="${CPU}" --memory="${MEMORY}" \
+    --image="${DOCKER_REGISTRY}/eq-questionnaire-launcher:${IMAGE_TAG}" --platform=managed --allow-unauthenticated \
```
Comment (ajmaddaford) [inline ci/deploy_app.sh:25]: As `DOCKER_REGISTRY` is required it should also be validated to ensure it exists.

**Review by MebinAbraham — APPROVED** [review]

File: ci/deploy.yaml [hunk 9]
```diff
@@ -2,15 +2,17 @@ platform: linux
 image_resource:
   type: docker-image
   source:
-    repository: ((image_registry))/eq-app-deploy-image
-    tag: ((deploy_image_version))
+    repository: gcr.io/google.com/cloudsdktool/cloud-sdk
+    tag: latest
```
Comment (MebinAbraham) [inline ci/deploy.yaml:6]: The full image is not needed. The `slim` version should be used.
```suggestion
    tag: slim
```

File: ci/deploy_app.sh (hunk 8 above)
Comment (petechd) [reply ci/deploy_app.sh:30]: Done, thanks!

**Review by ajmaddaford — APPROVED** [review]

**Pull request merged**